### ISSUE TITLE
Pixel Blenders not producing correct output

### ISSUE THREAD

**Issue by tocsoft** [SixLabors/ImageSharp#429]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of ImageSharp
- [x] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [x] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

Currently the Porter duff functions are producing incorrect outputs when the source image has transparent pixels.

For example taking  these images
#### Mask (blue fill with a transparent hole)
![mask](https://user-images.githubusercontent.com/166440/34760908-89de1b46-f5d9-11e7-8b58-a1c2c3b92037.png)

#### Source (full red fill)
![src](https://user-images.githubusercontent.com/166440/34760724-db213994-f5d8-11e7-96b4-432c142c5bd1.png)

And applying blending mode of `In` which should map to `Dest In` 

Should produce an output like this
#### Expected output (red fill with transparent hole)
![destin-expected](https://user-images.githubusercontent.com/166440/34760790-11cbbf8c-f5d9-11e7-9714-aa28ddd06615.png)

### Actual output (fully transparent image)
![destin](https://user-images.githubusercontent.com/166440/34760791-11e405d8-f5d9-11e7-9b4a-c5f8588d1d17.png)

### Steps to Reproduce

```c#
 Directory.CreateDirectory("output");
            Directory.CreateDirectory("output/src-mask");
            Directory.CreateDirectory("output/mask-src");
            using (var src = Image.Load("src.png"))
            using (var mask = Image.Load("mask.png"))
            {
                var modes = Enum.GetValues(typeof(SixLabors.ImageSharp.PixelFormats.PixelBlenderMode)).Cast<SixLabors.ImageSharp.PixelFormats.PixelBlenderMode>();
                foreach (var m in modes)
                {
                    using (var res = src.Clone(x => x.Blend(mask, new GraphicsOptions() { BlenderMode = m })))
                    {
                        res.Save($"output/src-mask/{m}.png");
                    }
                    using (var res = mask.Clone(x => x.Blend(src, new GraphicsOptions() { BlenderMode = m })))
                    {
                        res.Save($"output/mask-src/{m}.png");
                    }
                }
            }
```

### System Configuration

- ImageSharp version:
- Other ImageSharp packages and versions: master
- Environment (Operating system, version and so on): windows
- .NET Framework version: core2.0
- Additional information: n/a

**Comment by tocsoft** [comment]
We need to make sure we are fully compliant with the css compositing spec https://www.w3.org/TR/compositing-1/

**Comment by JimBobSquarePants** [comment]
Bugger... Is it just that method?

**Comment by tocsoft** [comment]
No its anywhere using the pixel blenders... but its most noticeable in that method as we are doing whole buffer effects.

Also don't think its all of the effects just a couple of them but I need to investigate some more to figure out the full scope of the issue.

**Comment by JimBobSquarePants** [comment]
Ok, fingers crossed it's something simple. The equations are all in the spec for comparison as I recall.

**Comment by JimBobSquarePants** [comment]
@tocsoft I've not looked at the formulas but these lines looks incredible suspicious to me. 

https://github.com/SixLabors/ImageSharp/blob/8899f23c1ddf8044d4dea7d5055386f684120761/src/ImageSharp/PixelFormats/PixelBlenders/PorterDuffFunctions.Generated.cs#L81

https://github.com/SixLabors/ImageSharp/blob/8899f23c1ddf8044d4dea7d5055386f684120761/src/ImageSharp/PixelFormats/PixelBlenders/PorterDuffFunctions.Generated.cs#L184

**Comment by JimBobSquarePants** [comment]
Ok.. So I created a small program based on the above code using images similar to the ones in the spec so we can compare

https://www.w3.org/TR/compositing-1/#porterduffcompositingoperators

[porter-duff-tests.zip](https://github.com/SixLabors/ImageSharp/files/1823911/porter-duff-tests.zip)

There's a few issues I've spotted straight away: 

- `DestIn`, and `In` are both definitely incorrect
- `Subtract` is wrong also (and misnamed)

It's difficult to know what is right or wrong though as we use different naming conventions from the spec.

We need to:

- Match naming
- Complete the list so we contain all the options.
- Fix our issues.

This is not something I'm confident in my own ability to solve so I will definitely need help.
